metrics_api/lib/opentelemetry/metrics/meter.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Metrics
+    # No-op implementation of Meter.
+    class Meter
+      COUNTER = Instrument::Counter.new
+      OBSERVABLE_COUNTER = Instrument::ObservableCounter.new
+      HISTOGRAM = Instrument::Histogram.new
+      OBSERVABLE_GAUGE = Instrument::ObservableGauge.new
+      UP_DOWN_COUNTER = Instrument::UpDownCounter.new
+      OBSERVABLE_UP_DOWN_COUNTER = Instrument::ObservableUpDownCounter.new
+
+      private_constant(:COUNTER, :OBSERVABLE_COUNTER, :HISTOGRAM, :OBSERVABLE_GAUGE, :UP_DOWN_COUNTER, :OBSERVABLE_UP_DOWN_COUNTER)
+
+      DuplicateInstrumentError = Class.new(OpenTelemetry::Error)
+
+      def initialize
+        @mutex = Mutex.new
+        @registry = {}
+      end
+
+      def create_counter(name, unit: nil, description: nil)
+        create_instrument(:counter, name, unit, description, nil) { COUNTER }
+      end
+
+      def create_observable_counter(name, unit: nil, description: nil, callback:)
+        create_instrument(:observable_counter, name, unit, description, callback) { OBSERVABLE_COUNTER }
+      end
+
+      def create_histogram(name, unit: nil, description: nil)
+        create_instrument(:histogram, name, unit, description, nil) { HISTOGRAM }
+      end
+
+      def create_observable_gauge(name, unit: nil, description: nil, callback:)
+        create_instrument(:observable_gauge, name, unit, description, callback) { OBSERVABLE_GAUGE }
+      end
+
+      def create_up_down_counter(name, unit: nil, description: nil)
+        create_instrument(:up_down_counter, name, unit, description, nil) { UP_DOWN_COUNTER }
+      end
+
+      def create_observable_up_down_counter(name, unit: nil, description: nil, callback:)
+        create_instrument(:observable_up_down_counter, name, unit, description, callback) { OBSERVABLE_UP_DOWN_COUNTER }
+      end
+
+      private
+
+      def create_instrument(kind, name, unit, description, callback)
```

Comment:
This should enforce [instrument naming requirements](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument) from the spec.

> Instrument names MUST conform to the following syntax (described using the
> [Augmented Backus-Naur Form](https://tools.ietf.org/html/rfc5234)):
> 
> ```abnf
> instrument-name = ALPHA 0*62 ("_" / "." / "-" / ALPHA / DIGIT)
> 
> ALPHA = %x41-5A / %x61-7A; A-Z / a-z
> DIGIT = %x30-39 ; 0-9
> ```
> 
> * They are not null or empty strings.
> * They are case-insensitive, ASCII strings.
> * The first character must be an alphabetic character.
> * Subsequent characters must belong to the alphanumeric characters, '_', '.',
>   and '-'.
> * They can have a maximum length of 63 characters.
> 

Additional requirements:

- `unit` must be ASCII with max length of 63 characters
- `description` must be `utf8mb3` and max length of 1023 characters

Reply:
https://github.com/open-telemetry/opentelemetry-ruby/pull/1004/commits/a9129ba1d87ad32914cf60d8795d9278d66718da